clients/src/main/java/org/apache/kafka/common/requests/FetchResponse.java

```diff
@@ -80,14 +89,26 @@ public Errors error() {
         return Errors.forCode(data.errorCode());
     }
 
-    public LinkedHashMap<TopicPartition, FetchResponseData.PartitionData> responseData() {
+    public LinkedHashMap<TopicPartition, FetchResponseData.PartitionData> responseData(Map<Uuid, String> topicNames, short version) {
+        return toResponseDataMap(topicNames, version);
+
+    }
+
+    // TODO: Should be replaced or cleaned up. The idea is that in KafkaApis we need to reconstruct responseData even though we could have just passed in and out a map.
+    //  With topic IDs, recreating the map takes a little more time since we have to get the topic name from the topic ID to name map.
+    //  The refactor somewhat helps in KafkaApis where we already have the topic names, but we have to recompute the map using topic IDs instead of just returning what we have.
+    //  Can be replaced when we remove toMessage and change sizeOf as a part of KAFKA-12410.
+    // Used when we can guarantee responseData is populated with all possible partitions
+    // This occurs when we have a response version < 13 or we built the FetchResponse with
+    // responseDataMap as a parameter and we have the same topic IDs available.
+    public LinkedHashMap<TopicPartition, FetchResponseData.PartitionData> resolvedResponseData() {
         if (responseData == null) {
             synchronized (this) {
                 if (responseData == null) {
                     responseData = new LinkedHashMap<>();
                     data.responses().forEach(topicResponse ->
-                            topicResponse.partitions().forEach(partition ->
-                                    responseData.put(new TopicPartition(topicResponse.topic(), partition.partitionIndex()), partition))
+                        topicResponse.partitions().forEach(partition ->
+                                responseData.put(new TopicPartition(topicResponse.topic(), partition.partitionIndex()), partition))
```

Comment:
The difference is that we already have the topic names here. I added it for the optimization when we build the response data map again in KafkaApis (so we don't have to look up the topic ids again). But it is a little silly. One option is just to put that logic in the one place it is used instead of having such a method.